### ISSUE TITLE
A redraw is needed to display the first line of the text inserted into an empty buffer by :insert or :append from a mapping or a timer

### ISSUE THREAD

**Issue by zeertzjq** [vim/vim#9438]
### Steps to reproduce

1. Run `vim --clean`
2. Create mapping `:nnoremap <F2> <Cmd>append<CR>aaaaa<CR>bbbbb<CR>.<CR>`
3. Press `<F2>`
4. The `bbbbb` is displayed immediately, but the `aaaaa` is not displayed until the next redraw

### Expected behaviour

The `aaaaa` is displayed immediately

### Version of Vim

Vim 8.2.3890

### Environment

Operating system: Arch Linux
Terminal: kitty
Value of $TERM: xterm-kitty
Shell: fish 3.3.1

**Comment by zeertzjq** [comment]
I think this is a side-effect of patch 7.4.1970. Patch 7.4.1970 changed `ex_append` to call `appended_lines_mark()` with `1` as `lnum` when the buffer is empty, which in turn calls `changed_lines()` with `2` as `lnum`, and the first line is not marked as modified.

**Comment by zeertzjq** [comment]
This bug also applies to text inserted into an empty buffer by :insert or :append from a timer:
```vim
func Timer(timer)
insert
aaaaa
bbbbb
.
endfunc

call timer_start(0, 'Timer')
```

Source the file in Vim, and the bug appears.

**Comment by zeertzjq** [comment]
This bug also applies to text inserted into an empty buffer by :insert or :append from a timer:
```vim
func Timer(timer)
insert
aaaaa
bbbbb
..
endfunc

call timer_start(0, 'Timer')
```
Source the file in Vim, and the bug appears.

**Comment by zeertzjq** [comment]
I'm trying to fix this in #9439
